### PR TITLE
Adding new package: COSTA

### PR DESCRIPTION
**Pull request by kabicm**
[COSTA](https://github.com/eth-cscs/COSTA) is a communication-optimal, highly-optimised algorithm for redistributing and transposing matrices that are potentially distributed across multiple MPI processes. It implements scalapack routines for matrix scale & transpose operations (`sub(C) = alpha * sub(A)^T + beta * C`, provided by `pxtran(u)`) and data redistribution (`sub(C) = sub(A)`, provided by `pxgemr2d`). Unlike previous redistribution algorithms, COSTA will also propose the relabelling of MPI ranks that minimizes the data reshuffling cost, leaving to users to decide if they want to use it. 

What makes COSTA more general than scalapack routines is that it is not limited only to block-cyclic data distributions, but can deal with completely arbitrary and irregular matrix distributions and can be easily generalized for n-dimensional tensors. 

Thanks to its scalapack wrappers, scalapack users do not need to change their code in order to use COSTA: it is enough to link your library to COSTA before linking to scalapack and all `pxtran, pxtranu` and `pxgemr2d` routines will automatically be using the COSTA algorithm.

This work is published in the **Proceedings of the International Conference on High Performance Computing (ISC21)** and is available under the following links:
- **published version:** https://link.springer.com/chapter/10.1007/978-3-030-78713-4_12
- **arxiv preprint:** https://arxiv.org/abs/2106.06601

### CODE REVIEW

File: var/spack/repos/builtin/packages/costa/package.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    url = 'https://github.com/eth-cscs/COSTA/releases/download/v2.0/COSTA-v2.0.tar.gz'
+    git = 'https://github.com/eth-cscs/COSTA.git'
+
+    # note: The default archives produced with github do not have the archives
+    #       of the submodules.
+    version('master', branch='master', submodules=True)
+    version('2.0', sha256='ef283b904b1b77b7d0de401cbdc3d7850c77368f2b57249c6eaee3017794c4b8')
+
+    variant('scalapack', default=False, description='Build with ScaLAPACK API')
+    variant('openmp', default=True, description="Build with OpenMP support")
```
Comment (haampie) [inline var/spack/repos/builtin/packages/costa/package.py]: shouldn't this be passed to cmake?

Comment (kabicm) [reply var/spack/repos/builtin/packages/costa/package.py]: You are right! I removed the openmp variant, as it's always assumed in cmake as well. The scalapack should be passed to cmake already.

Comment (kabicm) [reply var/spack/repos/builtin/packages/costa/package.py]: does it seem stuck or is it still fine?

**Comment by haampie** [comment]
@spackbot run pipeline

**Review by haampie — APPROVED** [review]

**Pull request merged**